x/gamm/keeper/multihop_test.go

```diff
@@ -133,70 +189,135 @@ func (suite *KeeperTestSuite) TestBalancerPoolSimpleMultihopSwapExactAmountOut()
 						suite.NoError(err, "test: %v", test.name)
-						dec = dec.Mul(sp)
 					}
-					return dec
-				}()
+				}
+
+				nextTokenOut := test.param.tokenOut
+				for i := len(test.param.routes) - 1; i >= 0; i-- {
+					hop := test.param.routes[i]
+					tokenOut, err := keeper.SwapExactAmountOut(cacheCtx, suite.TestAccs[0], hop.PoolId, hop.TokenInDenom, sdk.NewInt(100000000), nextTokenOut)
+					suite.Require().NoError(err)
+					nextTokenOut = sdk.NewCoin(hop.TokenInDenom, tokenOut)
+				}
+				return nextTokenOut
+			}
+
+			if test.expectPass {
+				tokenInCalculatedAsSeparateSwaps := calcInAmountAsSeparateSwaps(poolDefaultSwapFee)
+				tokenInCalculatedAsSeparateSwapsWithoutFee := calcInAmountAsSeparateSwaps(poolZeroSwapFee)
 
+				spotPriceBefore := calcSpotPrice()
+				multihopTokenInAmount, err := keeper.MultihopSwapExactAmountOut(suite.Ctx, suite.TestAccs[0], test.param.routes, test.param.tokenInMaxAmount, test.param.tokenOut)
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				spotPriceAfter := calcSpotPrice()
 				// Ratio of the token out should be between the before spot price and after spot price.
 				// This is because the swap increases the spot price
-				sp := tokenInAmount.ToDec().Quo(test.param.tokenOut.Amount.ToDec())
-				fmt.Printf("spBefore %s, spAfter %s, sp actual %s\n", spotPriceBefore, spotPriceAfter, sp)
+				sp := multihopTokenInAmount.ToDec().Quo(test.param.tokenOut.Amount.ToDec())
 				suite.True(sp.GT(spotPriceBefore) && sp.LT(spotPriceAfter), "multi-hop spot price wrong, test: %v", test.name)
+
+				if test.reducedFeeApplied {
+					// we have 3 values:
+					// ---------------- tokenInCalculatedAsSeparateSwaps (with full fees)
+					//    diffA
+					// ---------------- multihopTokenInAmount (with half fees)
+					//    diffB
+					// ---------------- tokenInCalculatedAsSeparateSwapsWithoutFee
+					// here we want to test, that difference between this 3 values around 50% (fee is actually halved)
+					// we do not have exact 50% reduce due to amm math rounding
+					// playing with input values for this test can result in different discount %
+					// so lets check, that we have around 50% +-1% reduction
+					diffA := tokenInCalculatedAsSeparateSwaps.Amount.Sub(multihopTokenInAmount)
+					diffB := multihopTokenInAmount.Sub(tokenInCalculatedAsSeparateSwapsWithoutFee.Amount)
+					diffDistinctionPercent := diffA.Sub(diffB).Abs().ToDec().Quo(diffA.Add(diffB).ToDec())
+					suite.Require().True(diffDistinctionPercent.LT(sdk.MustNewDecFromStr("0.01")))
+				} else {
+					suite.Require().True(multihopTokenInAmount.Equal(tokenInCalculatedAsSeparateSwaps.Amount))
+				}
 			} else {
 				_, err := keeper.MultihopSwapExactAmountOut(suite.Ctx, suite.TestAccs[0], test.param.routes, test.param.tokenInMaxAmount, test.param.tokenOut)
 				suite.Error(err, "test: %v", test.name)
 			}
 		})
 	}
 }
+
+func (s *KeeperTestSuite) UpdatePoolSwapFee(ctx sdk.Context, poolId uint64, adjustedPoolSwapFee sdk.Dec) error {
```

Comment:
I don't think we need to make this function public right?

Reply:
fixed